src/mainboard/protectli/vault_jsl/mainboard.c

```diff
@@ -3,9 +3,156 @@
 #include <arch/mmio.h>
 #include <device/device.h>
 #include <pc80/i8254.h>
+#include <device/pci_ids.h>
+#include <smbios.h>
 #include <soc/iomap.h>
+#include <soc/pci_devs.h>
 #include <soc/ramstage.h>
 
+smbios_enclosure_type smbios_mainboard_enclosure_type(void)
+{
+	return SMBIOS_ENCLOSURE_LOW_PROFILE_DESKTOP;
+}
+
+u8 smbios_mainboard_feature_flags(void)
+{
+	return SMBIOS_FEATURE_FLAGS_HOSTING_BOARD | SMBIOS_FEATURE_FLAGS_REPLACEABLE;
+}
+
+const char *smbios_system_sku(void)
+{
+	return CONFIG_MAINBOARD_PART_NUMBER "-01";
+}
+
+const char *smbios_chassis_sku(void)
+{
+	return CONFIG_MAINBOARD_PART_NUMBER "-001";
+}
+
+ const char *smbios_chassis_version(void)
+{
+	return "1.2";
+}
+
+static int smbios_type41_write_ethernet(struct device *parent, int *handle,
+					unsigned long *current)
+{
+	struct device *eth;
+	static u8 instance = 0;
+
+	if (!parent || !parent->enabled)
+		return 0;
+
+	eth = pcidev_path_behind(parent->link_list, PCI_DEVFN(0, 0));
+	if (!eth || !eth->enabled)
+		return 0;
+
+	/* Only i226 Ehternet Controller */
+	if (eth->vendor != PCI_VID_INTEL || eth->device != 0x125c)
+		return 0;
+
+	return smbios_write_type41(
+		current, handle,
+		"Onboard - Ethernet",		/* name */
+		instance++,			/* instance */
+		0,				/* segment */
+		eth->bus->secondary,		/* bus */
+		PCI_SLOT(eth->path.pci.devfn),	/* device */
+		PCI_FUNC(eth->path.pci.devfn),	/* function */
+		SMBIOS_DEVICE_TYPE_ETHERNET);	/* device type */
+}
+
+static int smbios_type41_write_ethernet_behind_switch(struct device *root_port, int *handle,
+						      unsigned long *current)
+{
+	struct device *switch_dev;
+	struct device *child;
+	int len = 0;
+
+	if (!root_port || !root_port->enabled)
+		return 0;
+
+	switch_dev = pcidev_path_behind(root_port->link_list, PCI_DEVFN(0, 0));
+	if (!switch_dev || !switch_dev->enabled)
+		return 0;
+
+	/* Only ASMedia switch */
+	if (switch_dev->vendor != 0x1b21 || switch_dev->device != 0x2806)
+		return 0;
+
+	/* After upstream port there is also one downstream port */
+	switch_dev = pcidev_path_behind(switch_dev->link_list, PCI_DEVFN(0, 0));
+	if (!switch_dev || !switch_dev->enabled)
+		return 0;
+
+	/* Only ASMedia switch */
+	if (switch_dev->vendor != 0x1b21 || switch_dev->device != 0x2806)
+		return 0;
```

Comment:
Nit: this is valid, but it looks like bad conflict resolution or copy&paste error, someone may try to remove it in the future. Could you merge these checks into one? Something like this (plus wrapping):

```suggestion
	if (!switch_dev || !switch_dev->enabled || switch_dev->vendor != 0x1b21 || switch_dev->device != 0x2806)
		return 0;
```

Reply:
> Nit: this is valid, but it looks like bad conflict resolution or copy&paste error,

I actually made these two checks deliberately separate, It looks more readable to me like that.